userspace/amix/man.1
AMIX(1)                User Commands                AMIX(1)

NAME
       amix - A command-line utility for various tasks.

SYNOPSIS
       amix [OPTION]... [SUBCOMMAND]...

DESCRIPTION
       The amix command is a versatile command-line utility that provides various subcommands for performing different tasks. This man page describes the "fsearch" and "search" subcommands.

SUBCOMMANDS
       - fsearch       Search for a word or phrase in a file and display the first match.

       - search        Search for a word or phrase in a file and count the matches.

OPTIONS
       None currently supported.

USAGE
       To use amix subcommands, type the desired subcommand followed by its arguments.

amix-fsearch SUBCOMMAND OPTIONS
       The amix-fsearch command is used to search for a specific word or phrase in a text file and display the first match. It supports the following options:

       - file_path: The path to the file in which you want to search.

       - search_text: The word or phrase you want to search for.

       - case_sensitive: An optional flag (1 for case-sensitive, 0 for case-insensitive, default is case-insensitive).

EXAMPLES
       - To search for the word "example" in a file: amix fsearch file.txt example

       - To perform a case-sensitive search: amix fsearch file.txt example 1

amix-search SUBCOMMAND OPTIONS
       The amix-search command is used to search for a specific word or phrase in a text file and count the matches. It supports the following options:

       - file_path: The path to the file in which you want to search.

       - search_text: The word or phrase you want to search for.

       - case_sensitive: An optional flag (1 for case-sensitive, 0 for case-insensitive, default is case-insensitive).

AUTHOR
       The amix command was created by [Your Name].

COPYRIGHT
       This is free software; you can redistribute it and/or modify
       it under the terms of the GNU General Public License as published by
       the Free Software Foundation; either version 3 of the License, or
       (at your option) any later version.

       This program is distributed in the hope that it will be useful,
       but WITHOUT ANY WARRANTY; without even the implied warranty of
       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
       GNU General Public License for more details.

       You should have received a copy of the GNU General Public License
       along with this program; if not, write to the Free Software Foundation,
       Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

SEE ALSO
       More information on the amix command and its subcommands may be available in the system documentation.


docs/CHANGELOG.md
<h1 align="center">CHANGELOG<h1>

# Version ALPHA 

- 7/29/2023(1.1) Fixed delete file function so it now frees up space op the disk

- 7/30/2023(1.2) Added a changelog file, A contributions file, support and specifications concerning the non GPLv3 license, A README file, Started documentation of code  

- 8/1/2023(1.3) Added support for multi sector files, still buggy but slowly coming along  

# VERSION BETA

- 03/10/2023(2.0) Added support for FAT32/16/12 from fat_io_lin and organized files


ui/man.c
#include "../include/string.h"
#include "../include/printf.h"
#include "../include/display.h"
#include "../include/fat_filelib.h"
#include "../include/man.h"
#include "../include/keyboard.h"
#include "../include/vesa.h"
#include "../include/vesa_display.h"
#define MAX_CHUNKS 20
#define MAX_LINES_PER_CHUNK 16
void man_inspect()
{
    printf("Inspect Command\n");
    printf("==============\n");
    printf("Description:\n");
    printf("    The 'inspect' command is used to inspect various system information.\n");
    printf("\nSubcommands:\n");
    printf("    1. inspect err\n");
    printf("        - Displays error information.\n");
    printf("    2. inspect log\n");
    printf("        - Displays event log information.\n");
    printf("        - Options:\n");
    printf("            a. --all\n");
    printf("                - Displays all events in the log.\n");
    printf("            b. --find\n");
    printf("                - Searches for specific events based on criteria.\n");
    printf("                - Prompts the user for start and end times, error status,\n");
    printf("                  and function name.\n");
    printf("    3. inspect pci\n");
    printf("        - Displays PCI device information.\n");
    printf("        - Options:\n");
    printf("            a. --list\n");
    printf("                - Lists all PCI devices.\n");
    printf("            b. --find <device>\n");
    printf("                - Searches for a specific PCI device by name.\n");
    printf("    4. inspect stack\n");
    printf("        - Displays stack trace information (currently disabled).\n");
    printf("    5. inspect drivers\n");
    printf("        - Displays driver information.\n");
    printf("        - Options:\n");
    printf("            a. --all\n");
    printf("                - Lists all registered devices.\n");
    printf("            b. --live\n");
    printf("                - Lists all currently running devices.\n");
}
// For the 'pci' command
void man_pci() {
    printf("PCI Command\n");
    printf("===========\n");
    printf("Description:\n");
    printf("    The 'pci' command is used to interact with PCI devices.\n");
    printf("\nSubcommands:\n");
    printf("    1. pci list\n");
    printf("        - Lists all available PCI devices.\n");
    printf("    2. pci reg\n");
    printf("        - Lists all registered PCI devices.\n");
}

// For the 'write' command
void man_write() {
    printf("Write Command\n");
    printf("=============\n");
    printf("Description:\n");
    printf("    The 'write' command is used to write data to a file.\n");
    printf("\nUsage:\n");
    printf("    write <filename> <file extension>\n");
}

// For the 'read' command
void man_read() {
    printf("Read Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'read' command is used to read the contents of a file.\n");
    printf("\nUsage:\n");
    printf("    read <filename> <file extension>\n");
}

// For the 'ls' command
void man_ls() {
    printf("LS Command\n");
    printf("==========\n");
    printf("Description:\n");
    printf("    The 'ls' command is used to list files and directories in the current directory.\n");
    printf("\nUsage:\n");
    printf("    ls\n");
}

// For the 'rm' command
void man_rm() {
    printf("RM Command\n");
    printf("==========\n");
    printf("Description:\n");
    printf("    The 'rm' command is used to remove (delete) a file.\n");
    printf("\nUsage:\n");
    printf("    rm <filename>\n");
}

// For the 'cls' command
void man_cls() {
    printf("CLS Command\n");
    printf("===========\n");
    printf("Description:\n");
    printf("    The 'cls' command is used to clear the screen.\n");
    printf("\nUsage:\n");
    printf("    cls\n");
}

// For the 'bg' command
void man_bg() {
    printf("BG Command\n");
    printf("==========\n");
    printf("Description:\n");
    printf("    The 'bg' command is used to change the background colors or themes.\n");
    printf("\nUsage:\n");
    printf("    bg\n");
}

// For the 'logo' command
void man_logo() {
    printf("Logo Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'logo' command is used to display a logo on the screen.\n");
    printf("\nUsage:\n");
    printf("    logo\n");
}

// For the 'mkdir' command
void man_mkdir() {
    printf("Mkdir Command\n");
    printf("==============\n");
    printf("Description:\n");
    printf("    The 'mkdir' command is used to create a new directory.\n");
    printf("\nUsage:\n");
    printf("    mkdir <directory_name>\n");
}

// For the 'exit' command
void man_exit() {
    printf("Exit Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'exit' command is used to exit the application or system.\n");
    printf("\nUsage:\n");
    printf("    exit\n");
}

// For the 'set-xy' command
void man_set_xy() {
    printf("Set-xy Command\n");
    printf("==============\n");
    printf("Description:\n");
    printf("    The 'set-xy' command is used to set the screen coordinates (X and Y positions).\n");
    printf("\nUsage:\n");
    printf("    set-xy <x> <y>\n");
}

// For the 'song' command
void man_song() {
    printf("Song Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'song' command is used to play a song or music.\n");
    printf("\nUsage:\n");
    printf("    song\n");
}

// For the 'format' command
void man_format() {
    printf("Format Command\n");
    printf("==============\n");
    printf("Description:\n");
    printf("    The 'format' command is used to format a disk or drive.\n");
    printf("\nUsage:\n");
    printf("    format <disk_number> <disk_name>\n");
}

// For the 'table' command
void man_table() {
    printf("Table Command\n");
    printf("==============\n");
    printf("Description:\n");
    printf("    The 'table' command is used to display partition table information.\n");
    printf("\nUsage:\n");
    printf("    table\n");
}

// For the '3d' command
void man_3d() {
    printf("3D Command\n");
    printf("===========\n");
    printf("Description:\n");
    printf("    The '3d' command is used to enable 3D rendering or display 3D graphics.\n");
    printf("\nUsage:\n");
    printf("    3d\n");
}

// For the 'rle' command
void man_rle() {
    printf("RLE Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'rle' command is used for a specific purpose. Please refer to the documentation.\n");
    printf("\nUsage:\n");
    printf("    rle\n");
}

// For the 'fs' command
void man_fs() {
    printf("FS Command\n");
    printf("===========\n");
    printf("Description:\n");
    printf("    The 'fs' command is used for file system-related operations. Please refer to the documentation.\n");
    printf("\nUsage:\n");
    printf("    fs\n");
}

// For the 'exe' command
void man_exe() {
    printf("Exe Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'exe' command is used to execute a command or program.\n");
    printf("\nUsage:\n");
    printf("    exe <command_or_program_name>\n");
}

// For the 'sect' command
void man_sect() {
    printf("raw Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'raw' command is used for sector-related operations. Please refer to the documentation.\n");
    printf("\nUsage:\n");
    printf("    raw <argument> <argument>\n");

}
void man_raw() {
    printf("Raw Command\n");
    printf("===========\n");
    printf("Description:\n");
    printf("    The 'raw' command is used for low-level disk sector operations.\n");
    printf("\nSubcommands:\n");
    
    printf("1. raw read <drive ><lba>\n");
    printf("   - Read a sector from the disk.\n");
    printf("   - <drive>: The drive to read from to read from.\n");
    printf("   - <lba>: The logical block address to read from.\n");
    printf("2. raw write <lba> <data> <file size>\n");
    printf("   - Write data to a sector on the disk.\n");
    printf("   - <lba>: The logical block address to write to.\n");
    printf("   - <data>: The data to write to the sector.\n");
    printf("   - <file size>: The number of chars you think you will write.\n");
}
// For the 'edit' command
void man_edit() {
    printf("Edit Command\n");
    printf("============\n");
    printf("Description:\n");
    printf("    The 'edit' command is used to edit a file. Please refer to the documentation.\n");
    printf("\nUsage:\n");
    printf("    edit <filename>\n");
}

void man_main(const char* functionName) {
    if (strcmp(functionName, "pci") == 0) {
        man_pci();
    } else if (strcmp(functionName, "write") == 0) {
        man_write();
    } else if (strcmp(functionName, "read") == 0) {
        man_read();
    } else if (strcmp(functionName, "ls") == 0) {
        man_ls();
    } else if (strcmp(functionName, "rm") == 0) {
        man_rm();
    } else if (strcmp(functionName, "cls") == 0) {
        man_cls();
    } else if (strcmp(functionName, "bg") == 0) {
        man_bg();
    } else if (strcmp(functionName, "logo") == 0) {
        man_logo();
    } else if (strcmp(functionName, "mkdir") == 0) {
        man_mkdir();
    } else if (strcmp(functionName, "exit") == 0) {
        man_exit();
    } else if (strcmp(functionName, "set-xy") == 0) {
        man_set_xy();
    } else if (strcmp(functionName, "song") == 0) {
        man_song();
    } else if (strcmp(functionName, "format") == 0) {
        man_format();
    } else if (strcmp(functionName, "table") == 0) {
        man_table();
    } else if (strcmp(functionName, "3d") == 0) {
        man_3d();
    } else if (strcmp(functionName, "rle") == 0) {
        man_rle();
    } else if (strcmp(functionName, "fs") == 0) {
        man_fs();
    } else if (strcmp(functionName, "exe") == 0) {
        man_exe();
    } else if (strcmp(functionName, "sect") == 0) {
        man_sect();
    } else if (strcmp(functionName, "edit") == 0) {
        man_edit();
    }else if (strcmp(functionName,"raw") == 0)
    {
        man_raw();
    }
     else if(strcmp(functionName,"inspect") == 0){
        man_inspect();

    }else {
        char path[50] = "/man/";
        strcat(path,functionName);
        strcat(path,".1");
        FILE *f = fl_fopen(path,"rb");
        if(f == NULL)
        {
            printf("Couldn't open name page for %s\n",functionName);
            return -1;
        }
        char buffer[1000*4];
        memset(buffer,0,sizeof(buffer));
        fl_fread(buffer,sizeof(char),4000,f);
        // printf("%s\n",buffer);
        set_scroll_mode(2);
        printf("%s\n",buffer);
        // set_scroll_mode(0);
    //     // display_large_man_page(buffer,600);
    //       size_t max_chunk_size = 50;  // You can adjust the maximum chunk size as needed
    //         char chunks[MAX_CHUNKS][MAX_LINES_PER_CHUNK][max_chunk_size];
            
    //         int chunk_count = store_buffer_in_chunks(buffer, max_chunk_size, chunks);
            
    //         for (int i = 0; i < chunk_count; i++) {
    //             printf("Chunk %d:\n", i);
    //             for (int j = 0; j < MAX_LINES_PER_CHUNK; j++) {
    //                 if (chunks[i][j][0] != '\0') {
    //                     printf("%s\n", chunks[i][j]);
    //                 }
    //             }
    //             printf("\n");
    //  }
        set_scroll_mode(0);
        // printf("No manual page available for '%s'\n", functionName);
    }
}


int store_buffer_in_chunks(const char* input_buffer, size_t max_chunk_size, char (*chunks)[MAX_LINES_PER_CHUNK][max_chunk_size]) {
    const char* delimiter = "\n";  // Line delimiter

    int chunk_count = 0;
    size_t current_chunk_size = 0;

    char* buffer_copy = strdup(input_buffer);  // Create a copy of the input buffer
    char* token = strtok(buffer_copy, delimiter);
    
    while (token != NULL && chunk_count < MAX_CHUNKS) {
        size_t line_size = strlen(token);
        
        if (current_chunk_size + line_size + 1 <= max_chunk_size) {
            // Add this line to the current chunk
            strcpy((*chunks)[chunk_count][current_chunk_size], token);
            current_chunk_size += line_size + 1;  // Add 1 for the newline character
        } else {
            // Start a new chunk
            current_chunk_size = 0;
            chunk_count++;
        }
        
        token = strtok(NULL, delimiter);
    }

    kfree(buffer_copy);  // Free the copy of the input buffer
    return chunk_count + 1;  // Add 1 for the last chunk
}






int man_list()
{
    char *cmds[] = {"inspect","raw","edit","sect","exe","fs","rle","3d","table","format","song","set-xy","exit","mkdir","logo","bg","cls","rm","ls","read","write","pci"};
    int option  = printArrayWithSelection(cmds,22);
    man_main(cmds[option]);
}
